src/WebJobs.Extensions.DurableTask/EntityScheduler/Proxy/OrchestrationClientProxy.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs
+{
+    internal class OrchestrationClientProxy : IEntityProxyContext
+    {
+        private readonly IDurableOrchestrationClient client;
+
+        internal OrchestrationClientProxy(IDurableOrchestrationClient client)
+        {
+            this.client = client;
+        }
+
+        public Task CallAsync(EntityId entityId, string operationName, object operationInput)
+        {
+            throw new NotSupportedException();
+        }
+
+        public Task<TResult> CallAsync<TResult>(EntityId entityId, string operationName, object operationInput)
+        {
+            throw new NotSupportedException();
+        }
+
+        public void Signal(EntityId entityId, string operationName, object operationInput)
+        {
+            this.client.SignalEntityAsync(entityId, operationName, operationInput).GetAwaiter().GetResult();
```

Comment:
Do you write using delegate only for signals?
I do not understand well the execution with the return value when using a delegate.

Is it something like the following?

```csharp
await ctx.CreateEntityProxy<TestEntityClasses.ICounter>(async entityProxy =>
{
    entityProxy.Set(10);

    // need await
    var ret = await entityProxy.Get();
});
```

It is a very difficult problem. I examined Akka's typed actor implementation, but it seems that it is mostly determined by return type and compromised.

Reply:
A variant of this idea would be to allow a user calls a `void` method directly, and the we could internally store the internal `Task` generated by each call to `SignalEntityAsync`. After the user's function completes, we can await any not-yet-awaited tasks inside the extension code. This would prevent the thread-blocking problem while still ensuring that we do actually await the task at some point.

The downside of my idea is that users would not be able to directly handle any exceptions generated by the void methods. If customers do want to handle these exceptions, we could consider adding a `FlushAsync` method that they could call to await all the tasks we stored up until this point. It's not super-clean, but it's similar to the `ICollector` and `IAsyncCollector` support for output bindings. In the future we could also consider this approach for client-side batching of operations.

Reply:
Based on feedback, I thought that there is a limit to directly returning an instance of Entity Interface.

How about changing the return type to <code>EntityProxy&lt;ICounterEntity&gt;</code> and manipulating it via its properties and methods?

![image](https://user-images.githubusercontent.com/1356444/59415712-fac5fe00-8dfe-11e9-960c-c0ddf178ad5a.png)

It's like integrating ideas from @sebastianburckhardt and @cgillum

Reply:
I am quite happy with how the proxy works right now from within orchestrations and entities, so I would prefer to make changes *only* to the API on the client. 

I like @cgillums suggestion of implicitly awaiting all tasks at the end of the function, and supporting a flush. However, it does require a bit more explanation and users may be confused when exceptions are thrown from an "invisible" location.

Another idea is to just stick to a slight variation of the original SingalEntityAsync API:

```csharp
await client.SignalEntityAsync<ICounterEntity>(proxy => proxy.Set(50));
```

The point here is that all you can do with the proxy on the client is to signal (call is not supported on clients), so we may as well call it a signal.

I will try this out in a fork.

Reply:
I have tried this out (see shibayan/azure-functions-durable-extension#1). I think it works quite well.

I have also converted all the samples to use the syntax (see shibayan/azure-functions-durable-extension#2).

Reply:
Great work! It looks good to me.
My goal was to deal with entities safely. I think that was achieved.

I will try to merge your PR.

Reply:
I merged the @sebastianburckhardt implementation and added CreateEntityProxy overload.